feat(statistical-detectors): call breakpoint microservice for txns

mjq-sentry opened the pull request:
When a potential transaction regression is found via moving average, query its timeseries and use the breakpoint microservice to check if an actual regression is present.

mjq-sentry commented on src/sentry/tasks/statistical_detectors.py:
```diff
@@ -241,6 +260,31 @@ def _detect_transaction_trends(
+            "project_id": [project_id],
+            "project_objects": [Project.objects.get(id=project_id)],
+        }
+        results = metrics_performance.timeseries_query(
```
Ideally this would probably use `metrics_performance.bulk_timeseries_query`, but this should be fine for the internal release.

Zylphrex replied:
I'd start by batching multiple transactions in a single query first. The reason being we want to minimize the number of concurrent queries we make to clickhouse. So where possible, try to query multiple timeseries in a single query.

udameli replied:
I think we should still try minimizing the number of queries we're making here. You can maybe batch multiple transaction timeseries in a single query with a `Column("transaction")` groupBy.

Snuba allows 10,000 data points per request. 14 days * 1hr * 24hr = 336 data points per transaction name. We can safely get 25 transaction name timeseries

mjq-sentry commented:
This needs the same delay work that was added to the functions check

mjq-sentry commented:
Delay added to match functions!

Zylphrex commented on src/sentry/tasks/statistical_detectors.py:
```diff
@@ -241,6 +303,31 @@ def _detect_transaction_trends(
+        }
+        results = metrics_performance.timeseries_query(
+            selected_columns=[agg_function],
+            query=f"transaction:#{transaction_name}",
```
what is the `#` for? Also, I think you have to check if the transaction name needs to be escaped.

mjq-sentry replied:
Thanks! # is old Ruby muscle memory. Do you know if we have a function for escaping query args already kicking around? 🤔

mjq-sentry replied:
Couldn't find a conclusive answer in [the docs](https://docs.sentry.io/product/sentry-basics/search/) but I'm assuming that `"` is the only char needing escaping if you're in quotes. If so, should be escaped now.

udameli commented on src/sentry/tasks/statistical_detectors.py:
```diff
@@ -241,6 +303,31 @@ def _detect_transaction_trends(
+        }
+        results = metrics_performance.timeseries_query(
+            selected_columns=[agg_function],
+            query=f"transaction:#{transaction_name}",
```
not sure `#` is needed here since we don't do any special handling for it so it might just be interpreted as part of the transaction name

mjq-sentry replied:
Good catch! Sorry, Ruby muscle memory

Zylphrex reviewed (approved):
A few comments but overall flow looks good to me. Not sure if the data massaging is correct but I'll leave that to you to double check 👍

udameli commented on src/sentry/tasks/statistical_detectors.py:
```diff
@@ -241,6 +307,78 @@ def _detect_transaction_trends(
+                    params["end"],
+                    interval,
+                )
+                yield project_id, transaction_name, formatted_result
```
we'd still need to format each key as `{project_id},{transaction_name}` to match what the breakpoint microservice is expecting

mjq-sentry commented:
Just pushed again with a test that actually creates metrics, confirming that the query and all data munging is correct all the way through to issue creation. Added issue creation too, since it was effectively a one-liner and helped made sure everything up to that point was exercised.

Comments re: adding the TODO and the batch size are also addressed! 👍 

The only mystery remaining is that the query to retrieve multiple projects might only be retrieving one 🤔 In the test, the generated SNQL query when looking for `transaction_1` and `transaction_2` in the same project is:

```sql
MATCH (generic_metrics_distributions) SELECT arrayElement(quantilesIf(0.95)(value, equals(metric_id, 9223372036854775909)), 1) AS `p95_transaction_duration` BY transaction, toStartOfInterval(timestamp, toIntervalSecond(3600), 'Universal') AS `time` WHERE transform(tags_raw[9223372036854776020], array(''), array('<< unparameterized >>')) AS `transaction` = 'transaction_1' AND timestamp >= toDateTime('2023-09-19T11:00:00') AND timestamp < toDateTime('2023-10-03T11:00:00') AND project_id IN array(4552742136971281) AND org_id = 4552742136905744 AND metric_id IN array(9223372036854775909) ORDER BY toStartOfInterval(timestamp, toIntervalSecond(3600), 'Universal') AS `time` ASC LIMIT 10000 GRANULARITY 3600
```

Looks fine to me, but the results only incldue data from one of the transactions:

```python
{'data': [{'transaction': 'transaction_1', 'time': '2023-10-03T08:00:00+00:00', 'p95_transaction_duration': 1.0}, {'transaction': 'transaction_1', 'time': '2023-10-03T09:00:00+00:00', 'p95_transaction_duration': 9.5}], 'meta': [{'name': 'transaction', 'type': 'String'}, {'name': 'time', 'type': "DateTime('Universal')"}, {'name': 'p95_transaction_duration', 'type': 'Float64'}]}
```

Needs a bit more investigation.

Otherwise, functionally everything LGTM.
